trivy and cargo-audit for vulnerability scanning in CI

DaMandal0rian opened the pull request:
The PR adds vulnerability scanning for containers and rust packages using popular open-source tools:
- trivy container and repo scanner
- cargo-audit for crate scanning

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

nazar-pc commented on .github/workflows/rustsec-audit.yml:
```diff
@@ -0,0 +1,23 @@
+    - cron: "40 13 * * 0"
+jobs:
+  security_audit:
+    runs-on: ${{ fromJson(github.repository_owner == 'subspace' && '["self-hosted", "ubuntu-20.04-x86-64"]' || 'ubuntu-22.04') }}
```
This is incorrect syntax, this is the correct one:
```suggestion
    runs-on: ${{ fromJson(github.repository_owner == 'subspace' && '["self-hosted", "ubuntu-20.04-x86-64"]' || '"ubuntu-22.04"') }}
```

DaMandal0rian replied:
Thanks, fixed locally!

nazar-pc commented on .github/workflows/rustsec-audit.yml:
```diff
@@ -0,0 +1,23 @@
+    runs-on: ${{ fromJson(github.repository_owner == 'subspace' && '["self-hosted", "ubuntu-20.04-x86-64"]' || 'ubuntu-22.04') }}
+    steps:
+      - uses: actions/checkout@v3
+      - uses: rustsec/audit-check@v1.4.1
```
What about specific hashes like in all other actions in this repo?

DaMandal0rian replied:
This is already done too locally, was just doing this as a test first!

nazar-pc commented on .github/workflows/rustsec-audit.yml:
```diff
@@ -0,0 +1,23 @@
+      - uses: rustsec/audit-check@v1.4.1
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+        continue-on-error: true
```
Why?

DaMandal0rian replied:
This is needed for scan results to be published to Github security scanning https://github.com/subspace/subspace/pull/2378#issuecomment-1874864321

nazar-pc replied:
`continue-on-error: true` needed for scan results? I don't get it

DaMandal0rian replied:
It's not needed was left over from when I was testing the security advisories 🤦 
https://github.com/subspace/subspace/actions/runs/7624889136/job/20769203937

nazar-pc commented on .github/workflows/sec-scan.yml:
```diff
@@ -0,0 +1,199 @@
+##
```
I agree it would be good to scan container images, but I'm not a fan of the amount of copy-paste here and how much maintenance burden it will cause.

We either need to refactor container image building into reusable action first or even better, scan the actual latest container image we have already published, because even if fresh container image is not vulnerable, it says nothing about vulnerability of the latest published container image.

DaMandal0rian replied:
This wasn't copy-paste from the official docs, i was just doing this to test the tools. The code will be refactored to only use the scan jobs, not the whole container building process. Either include in our current workflows or use reusable workflows.

DaMandal0rian commented on .github/workflows/sec-scan.yml:
```diff
@@ -0,0 +1,199 @@
+          build-args: |
+            SUBSTRATE_CLI_GIT_COMMIT_HASH=${{ github.sha }}
+            RUSTFLAGS=${{ matrix.platform.rustflags }}
+
```
WIll be removed

DaMandal0rian commented on .github/workflows/sec-scan.yml:
```diff
@@ -0,0 +1,199 @@
+          # We build AArch64
+          - arch: linux/aarch64
+            dockerfile-suffix: ".aarch64"
+            suffix: ubuntu-aarch64-${{ github.ref_name }}
```
will be removed.

DaMandal0rian reviewed (commented):
This was a test that i didn't want to interfere with current workflows just to see how the scan tools work, hence why it's in draft still. Wasn't planning on using all that docker build and push stuff since we just need to use the latest images built on releases.

nazar-pc commented on .github/workflows/sec-scan.yml:
```diff
@@ -0,0 +1,60 @@
+      - name: Run Trivy vulnerability scanner on image
+        uses: aquasecurity/trivy-action@d43c1f16c00cfd3978dde6c07f4bbcf9eb6993ca # @v0.16.1
+        with:
+          image-ref: ghcr.io/${{ github.repository_owner }}/${{ matrix.image }}:${{ github.event.client_payload.tag }}
```
Why using tag (whose contents can change) rather than image hash?

DaMandal0rian replied:
True, omitted that and just sending the hash

nazar-pc commented on .github/workflows/snapshot-build.yml:
```diff
@@ -80,6 +80,12 @@ jobs:
+        uses: peter-evans/repository-dispatch@a4a90276d01e3a2ae44fa10a0247287f045afd59 # @v2.1.2
+        with:
+          event-type: sec-scan-trigger
+          client-payload: '{"ref": "${{ github.ref }}", "sha": "${{ github.sha }}" , "tag": "${{ steps.meta.outputs.tags }}" }'
```
We only seem to use `tag` on the receiving side, why sending so much stuff?

Also `tags` seems to imply multiple values, but you're using it as a single value instead.

DaMandal0rian replied:
changed this to just send that sha hash and use that instead.

nazar-pc commented on .github/workflows/sec-scan.yml:
I think this file should be named more descriptive than `sec-scan.yml`

nazar-pc commented on .github/workflows/sec-scan.yml:
```diff
@@ -0,0 +1,60 @@
+          severity: CRITICAL
+
+      - name: Upload Trivy scan results to GitHub Security tab
+        uses: github/codeql-action/upload-sarif@v2
```
No hash again

DaMandal0rian replied:
thanks, that has been corrected.

nazar-pc commented on .github/workflows/snapshot-build.yml:
```diff
@@ -80,6 +80,12 @@ jobs:
             SUBSTRATE_CLI_GIT_COMMIT_HASH=${{ github.sha }}
             RUSTFLAGS=${{ matrix.platform.rustflags }}
 
+      - name: Trigger sec-scan Workflow
```
Will block the workflow, or just trigger another one and move on?

DaMandal0rian replied:
It shouldn't it's triggering a webhook to dispatch it to the other workflow from what I see.

nazar-pc commented on .github/workflows/sec-scan.yml:
```diff
@@ -0,0 +1,60 @@
+
+on:
+  repository_dispatch:
+    types: [sec-scan-trigger]
```
This should also be called more descriptively

DaMandal0rian replied:
Changed it to `trivy-scan-dispatch`

nazar-pc reviewed (commented):
I squashed unnecessary commits further. The new implementation is much cleaner, but I have some additional comments with improvements.

DaMandal0rian reviewed (commented):
Thanks for the feedback!

DaMandal0rian committed:
add trivy, anchore and cargo-audit scans

- Fix naming and use sha hashes for packages
- minor fixes
- use hash instead of tags

nazar-pc commented on .github/workflows/trivy-security-scan.yml:
```diff
@@ -0,0 +1,60 @@
+          ignore-unfixed: true
+          format: sarif
+          output: trivy-results.sarif
+          severity: CRITICAL
```
Why no high here like above?
